vta/hardware/chisel/src/main/scala/Compute.scala

```diff
@@ -0,0 +1,364 @@
+class OutQueue[T <: Data](gen: T, entries: Int) extends Queue (gen, entries) {}
+
+class Compute(implicit val p: Parameters) extends Module with CoreParams {
+  val io = IO(new ComputeIO)
+
+  io.inp_mem <> DontCare
+  io.wgt_mem <> DontCare
+
+  val started = !reset.toBool
+
+  val acc_mem = Mem(1 << 8, UInt(512.W))
+  val uop_mem = Mem(1 << 10, UInt(32.W))
+
+  val insn            = Reg(UInt(128.W))
+  val insn_valid      = (insn =/= 0.U) && started
+
+  val opcode          = insn(insn_mem_0_1, insn_mem_0_0)
+  val pop_prev_dep    = insn(insn_mem_1)
+  val pop_next_dep    = insn(insn_mem_2)
+  val push_prev_dep   = insn(insn_mem_3)
+  val push_next_dep   = insn(insn_mem_4)
+
+  val memory_type = insn(insn_mem_5_1, insn_mem_5_0)
+  val sram_base   = insn(insn_mem_6_1, insn_mem_6_0)
+  val dram_base   = insn(insn_mem_7_1, insn_mem_7_0)
+  val y_size      = insn(insn_mem_8_1, insn_mem_8_0)
+  val x_size      = insn(insn_mem_9_1, insn_mem_9_0)
+  val x_stride    = insn(insn_mem_a_1, insn_mem_a_0)
+  val y_pad_0     = insn(insn_mem_b_1, insn_mem_b_0)
+  val y_pad_1     = insn(insn_mem_c_1, insn_mem_c_0)
+  val x_pad_0     = insn(insn_mem_d_1, insn_mem_d_0)
+  val x_pad_1     = insn(insn_mem_e_1, insn_mem_e_0)
+
+  // write to out_mem
+  val uop_bgn = Reg(UInt(16.W))
+  uop_bgn := insn(insn_gem_6_1, insn_gem_6_0)
+  val uop_end = Reg(UInt(16.W))
+  uop_end := insn(insn_gem_7_1, insn_gem_7_0)
+  val iter_out = insn(insn_gem_8_1, insn_gem_8_0)
+  val iter_in = insn(insn_gem_9_1, insn_gem_9_0)
+  val dst_factor_out = 1.U(16.W)
+  val dst_factor_in  = 1.U(16.W)
+  val src_factor_out = 1.U(16.W)
+  val src_factor_in  = 1.U(16.W)
+  val wgt_factor_out = 1.U(16.W)
+  val wgt_factor_in  = 1.U(16.W)
+  val alu_opcode = insn(insn_alu_e_1, insn_alu_e_0)
+  val use_imm = insn(insn_alu_f)
+  val imm_raw = RegNext(insn(insn_alu_g_1, insn_alu_g_0))
+  val imm = RegNext(Mux(imm_raw.asSInt < 0.S, Cat("hffff".U, imm_raw), Cat("h0000".U, imm_raw)).asSInt)
+
+  val sram_idx = sram_base
+  val dram_idx = dram_base
+  val y_size_total = y_pad_0 + y_size + y_pad_1
+  val x_size_total = x_pad_0 + x_size + x_pad_1
+  val y_offset = x_size_total * y_pad_0
+
+  // control and status registers (csr)
+  val opcode_finish_en = (opcode === opcode_finish.U)
+  val opcode_load_en = (opcode === opcode_load.U || opcode === opcode_store.U)
+  val opcode_gemm_en = (opcode === opcode_gemm.U)
+  val opcode_alu_en = (opcode === opcode_alu.U)
+
+  val memory_type_uop_en = memory_type === mem_id_uop.U
+  val memory_type_acc_en = memory_type === mem_id_acc.U
+
+  // status
+  val state = RegInit(0.U(3.W))
+  val s_IDLE :: s_DUMP :: s_BUSY :: s_PUSH :: s_DONE :: Nil = Enum(5)
+  val idle = state === s_IDLE
+  val dump = state === s_DUMP
+  val busy = state === s_BUSY
+  val push = state === s_PUSH
+  val done = state === s_DONE
+
+  // uops / biases / out_mem
+  val uops_read   = Reg(Bool())
+  val uops_data   = Reg(UInt((32 + 128).W))
```

Comment:
Is there an explanation for these magic numbers? Looks like maybe `(insn.getWidth + uop_mem.getWidth).W`. Similarly with the rest of the magic numbers in this section.